mmdet3d/datasets/pipelines/transforms_3d.py

```diff
@@ -650,6 +650,288 @@ def __repr__(self):
+            s_y = e_y - self.block_size
+            for idx_x in range(num_grid_x):
+                s_x = coord_min[0] + idx_x * stride
+                e_x = min(s_x + self.block_size, coord_max[0])
+                s_x = e_x - self.block_size
+                # boundary of a patch
+                cur_min = np.array([s_x, s_y, coord_min[2]])
+                cur_max = np.array([e_x, e_y, coord_max[2]])
+                cur_choice = np.sum(
+                    (coords >= (cur_min - 0.001)) * (coords <=
+                                                     (cur_max + 0.001)),
+                    axis=1) == 3
+
+                if not cur_choice.any():  # no points in this patch
+                    continue
+
+                # sample points in this patch to multiple batches
+                cur_center = cur_min + self.block_size / 2.0
+                point_idxs = np.where(cur_choice)[0]
+                num_batch = int(
+                    np.ceil(point_idxs.shape[0] / float(self.num_points)))
+                point_size = int(num_batch * self.num_points)
+                replace = point_size > 2 * point_idxs.shape[0]
+                point_idxs_repeat = np.random.choice(
+                    point_idxs,
+                    point_size - point_idxs.shape[0],
+                    replace=replace)
+                choices = np.concatenate([point_idxs, point_idxs_repeat])
+                np.random.shuffle(choices)
+
+                # construct model input
+                point_batch = self._input_generation(coords[choices],
+                                                     cur_center, coord_max,
+                                                     attributes[choices],
+                                                     attribute_dims,
+                                                     point_type)
+
+                all_points.append(point_batch)
+                all_idxs.append(choices)
+
+        all_points = point_type.cat(all_points)
+        all_idxs = np.concatenate(all_idxs)
+
+        # make sure all points are sampled at least once
+        assert np.unique(all_idxs).shape[0] == points.shape[0]
+
+        return all_points
+
+    def __call__(self, results):
+        """Call function to sample points to in indoor scenes.
+
+        Args:
+            input_dict (dict): Result dict from loading pipeline.
+
+        Returns:
+            dict: Results after sampling, 'points', 'pts_instance_mask' \
+                and 'pts_semantic_mask' keys are updated in the result dict.
+        """
+        points = results['points']
+
+        if self.test_mode:
+            points = self._sliding_patch_generation(points)
+            results['points'] = points
+            return results
+
+        pts_semantic_mask = results['pts_semantic_mask']
+
+        points, choices = self._patch_points_sampling(points,
+                                                      pts_semantic_mask)
```

Comment:
If pts_semantic_mask is None, is there an exception in method `_patch_points_sampling` (`cur_sem_mask = sem_mask[cur_choice]`)?

Reply:
Since test_mode is specified in this function, I raise an exception if no semantic mask is provided.